x-pack/plugins/lens/public/xy_visualization/xy_config_panel.tsx

```diff
@@ -118,14 +121,114 @@ export function LayerContextMenu(props: VisualizationLayerWidgetProps<State>) {
 }
 
 export function XyToolbar(props: VisualizationToolbarProps<State>) {
+  const axes = [
+    {
+      id: 'x',
+      label: 'X-axis',
+    },
+    {
+      id: 'y',
+      label: 'Y-axis',
+    },
+  ];
+
+  const { frame, state, setState } = props;
+
   const [open, setOpen] = useState(false);
-  const hasNonBarSeries = props.state?.layers.some(
+  const hasNonBarSeries = state?.layers.some(
     (layer) => layer.seriesType === 'line' || layer.seriesType === 'area'
   );
+
+  const [xaxistitle, setXaxistitle] = useState(state?.xTitle);
+  const [yaxistitle, setYaxistitle] = useState(state?.yTitle);
+
+  const xyTitles = useCallback(() => {
+    const defaults = {
+      xTitle: state?.xTitle,
+      yTitle: state?.yTitle,
+    };
+    const layer = state?.layers[0];
+    if (!layer || !layer.accessors.length) {
+      return defaults;
+    }
+    const datasource = frame.datasourceLayers[layer.layerId];
+    if (!datasource) {
+      return defaults;
+    }
+    const x = layer.xAccessor ? datasource.getOperationForColumnId(layer.xAccessor) : null;
+    const y = layer.accessors[0] ? datasource.getOperationForColumnId(layer.accessors[0]) : null;
+
+    return {
+      xTitle: defaults.xTitle || x?.label,
+      yTitle: defaults.yTitle || y?.label,
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [open]);
+
+  useEffect(() => {
+    const {
+      xTitle,
+      yTitle,
+    }: { xTitle: string | undefined; yTitle: string | undefined } = xyTitles();
+    setXaxistitle(xTitle);
+    setYaxistitle(yTitle);
+  }, [xyTitles]);
+
+  const [tickLabelsVisibilitySettings, setTickLabelsVisibilitySettings] = useState({
```

Comment:
Why are you storing all of these settings both in local state and in the frame state (via `setState`)? When you call `setState`, the frame will re-render the toolbar and value will be adjusted. We should try to avoid local state whenever possible.

Reply:
If I remember correctly, without localState is extremely slow, have to check it again though

Reply:
That's a good point, and an interesting solution.

For the buttons it should be irrelevant, for the label in the dimension popover we used this workaround (basically debouncing the state update): https://github.com/elastic/kibana/blob/b1655384866ef84aaf391f00924f109a838b0fef/x-pack/plugins/lens/public/indexpattern_datasource/dimension_panel/popover_editor.tsx#L59

It keeps the ugly performance optimization in a single place, so the top level form component doesn't have to deal with it.

Reply:
I can totally remove the local state for tickLabels and gridlines :)